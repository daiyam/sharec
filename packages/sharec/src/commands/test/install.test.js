const path = require('path')
const { readFileSync } = require.requireActual('fs')
const { vol } = require('memfs')
const install = require('../install')

describe('commands > install >', () => {
  const packageJson01 = require('fixtures/package/package_01.json')
  const packageJson02 = require('fixtures/package/package_02.json')
  const babel01 = require('fixtures/babel/json/babel_01.json')
  const babel02 = require('fixtures/babel/json/babel_02.json')
  const eslint01 = require('fixtures/eslint/json/eslintrc_01.json')
  const eslint02 = require('fixtures/eslint/json/eslintrc_02.json')
  const yamlEslint01 = readFileSync(
    path.resolve(
      __dirname,
      '../../../test/fixtures/eslint/yaml/eslintrc_01.yml',
    ),
    'utf8',
  )
  const yamlEslint02 = readFileSync(
    path.resolve(
      __dirname,
      '../../../test/fixtures/eslint/yaml/eslintrc_02.yml',
    ),
    'utf8',
  )

  beforeEach(() => {
    vol.reset()
  })

  describe('configuration processing', () => {
    describe('invalid cases', () => {
      it('should stops if configuration directory is not exists', async () => {
        expect.assertions(1)

        const packageJson = {}
        const dir = {
          '/target/package.json': JSON.stringify(packageJson),
          '/configuration-package/.editorconfig': 'bar',
        }

        vol.fromJSON(dir, '/')

        await install({
          configsPath: '/configuration-package',
          targetPath: '/target',
        })

        expect(
          JSON.parse(vol.readFileSync('/target/package.json', 'utf8')),
        ).toEqual(packageJson)
      })

      it('should not do anything if config already injected', async () => {
        expect.assertions(1)

        const packageJson = {
          sharec: {
            injected: true,
          },
        }
        const dir = {
          '/target/package.json': JSON.stringify(packageJson, null, 2),
          '/configuration-package/configs/package.json': JSON.stringify(
            packageJson01,
            null,
            2,
          ),
        }

        vol.fromJSON(dir, '/')

        await install({
          configsPath: '/configuration-package',
          targetPath: '/target',
        })

        expect(
          JSON.parse(vol.readFileSync('/target/package.json', 'utf8')),
        ).toEqual(packageJson)
      })
    })

    describe('valid cases', () => {
      it('should merge all configs from target path', async () => {
        expect.assertions(6)

        const dir = {
          '/target/.eslintrc': JSON.stringify(eslint01),
          '/target/babelrc.js': JSON.stringify(babel01),
          '/target/.eslintrc.yaml': yamlEslint01,
          '/target/package.json': JSON.stringify(packageJson01, null, 2),
          '/configuration-package/configs/.eslintrc': JSON.stringify(eslint02),
          '/configuration-package/configs/.eslintrc.yaml': yamlEslint02,
          '/configuration-package/configs/.editorconfig': 'bar',
          '/configuration-package/configs/.foo': JSON.stringify({
            foo: 'bar',
            bar: {
              foo: 'bar',
              bar: ['foo', 'bar'],
            },
          }),
          '/configuration-package/configs/babelrc.js': JSON.stringify(babel02),
          '/configuration-package/configs/package.json': JSON.stringify(
            packageJson02,
            null,
            2,
          ),
        }

        vol.fromJSON(dir, '/')

        await install({
          configsPath: '/configuration-package',
          targetPath: '/target',
        })

        expect(
          vol.readFileSync('/target/.editorconfig', 'utf8'),
        ).toMatchSnapshot()
        expect(vol.readFileSync('/target/babelrc.js', 'utf8')).toMatchSnapshot()
        expect(
          JSON.parse(vol.readFileSync('/target/.eslintrc', 'utf8')),
        ).toMatchSnapshot()
        expect(
          vol.readFileSync('/target/.eslintrc.yaml', 'utf8'),
        ).toMatchSnapshot()
        expect(
          vol.readFileSync('/target/package.json', 'utf8'),
        ).toMatchSnapshot()
        expect(
          JSON.parse(vol.readFileSync('/target/.foo', 'utf8')),
        ).toMatchSnapshot()
      })

      //       it('should backup origin configuration to sharec-lock.json file', async () => {
      //         expect.assertions(1)
      //
      //         const dir = {
      //           '/target/.eslintrc': JSON.stringify(eslint01),
      //           '/target/babelrc.js': JSON.stringify(babel01),
      //           '/target/.eslintrc.yaml': yamlEslint01,
      //           '/target/package.json': JSON.stringify(packageJson01, null, 2),
      //           '/configuration-package/configs/.eslintrc': JSON.stringify(eslint02),
      //           '/configuration-package/configs/.eslintrc.yaml': yamlEslint02,
      //           '/configuration-package/configs/.editorconfig': 'bar',
      //           '/configuration-package/configs/babelrc.js': JSON.stringify(babel02),
      //           '/configuration-package/configs/package.json': JSON.stringify(
      //             packageJson02,
      //             null,
      //             2,
      //           ),
      //         }
      //
      //         vol.fromJSON(dir, '/')
      //
      //         await install({
      //           configsPath: '/configuration-package',
      //           targetPath: '/target',
      //         })
      //
      //         expect(
      //           vol.readFileSync('/target/sharec-lock.json', 'utf8'),
      //         ).toMatchSnapshot()
      //       })
    })
  })
})
