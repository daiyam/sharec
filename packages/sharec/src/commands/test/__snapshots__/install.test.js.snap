// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commands > install > configuration processing valid cases should merge all configs from target path 1`] = `"bar"`;

exports[`commands > install > configuration processing valid cases should merge all configs from target path 2`] = `
"{
  \\"ignore\\": [
    \\"*.babelrc\\"
  ],
  \\"presets\\": [
    [
      \\"env\\",
      {
        \\"modules\\": false,
        \\"targets\\": {
          \\"browsers\\": [
            \\"> 1%\\",
            \\"last 2 versions\\",
            \\"not ie > 0\\"
          ]
        }
      }
    ],
    \\"stage-2\\"
  ],
  \\"plugins\\": [
    [
      \\"syntax-dynamic-import\\",
      {
        \\"foo\\": \\"bar\\"
      }
    ],
    \\"lodash\\"
  ],
  \\"env\\": {
    \\"test\\": {
      \\"presets\\": [
        \\"env\\",
        \\"stage-2\\"
      ],
      \\"plugins\\": [
        \\"dynamic-import-node\\"
      ]
    }
  }
}"
`;

exports[`commands > install > configuration processing valid cases should merge all configs from target path 3`] = `
Object {
  "env": Object {
    "browser": true,
    "es6": true,
    "jest": true,
    "node": true,
  },
  "extends": "airbnb",
  "globals": Object {
    "bar": true,
    "foo": true,
  },
  "parserOptions": Object {
    "ecmaFeatures": Object {
      "spread": true,
    },
    "ecmaVersion": 7,
    "parser": "babel-eslint",
    "sourceType": "module",
  },
  "rules": Object {
    "max-len": Array [
      "warn",
      120,
      4,
      Object {
        "ignoreComments": true,
        "ignoreStrings": false,
        "ignoreUrls": false,
      },
    ],
    "no-console": 0,
    "space-before-function-paren": 1,
  },
}
`;

exports[`commands > install > configuration processing valid cases should merge all configs from target path 4`] = `
"env:
  browser: true
  es6: true
  node: true
  jest: true
globals:
  foo: true
  bar: true
extends: airbnb
parserOptions:
  parser: babel-eslint
  ecmaVersion: 7
  ecmaFeatures:
    spread: true
  sourceType: module
rules:
  space-before-function-paren: 1
  no-console: 0
  max-len:
    - warn
    - 120
    - 4
    - ignoreUrls: false
      ignoreComments: true
      ignoreStrings: false
"
`;

exports[`commands > install > configuration processing valid cases should merge all configs from target path 5`] = `
"{
  \\"name\\": \\"my-awesome-package\\",
  \\"scripts\\": {
    \\"commit\\": \\"\\",
    \\"lint\\": \\"eslint 'src/**/*.{js,jsx,json}'\\",
    \\"lint-staged\\": \\"lint-staged\\"
  },
  \\"dependencies\\": {
    \\"react\\": \\"16.0.0\\"
  },
  \\"lint-staged\\": {
    \\"src/**/*.{js,jsx,json}\\": [
      \\"eslint --fix\\",
      \\"prettier --write\\",
      \\"git add\\"
    ]
  },
  \\"husky\\": {
    \\"hooks\\": {
      \\"pre-commit\\": \\"lint-staged\\",
      \\"commit-msg\\": \\"commitlint -E HUSKY_GIT_PARAMS\\"
    }
  },
  \\"config\\": {
    \\"commitizen\\": {
      \\"path\\": \\"cz-conventional-changelog\\"
    }
  },
  \\"prettier\\": {
    \\"semi\\": false,
    \\"singleQuote\\": true,
    \\"trailingComma\\": \\"all\\"
  },
  \\"eslintConfig\\": {
    \\"parser\\": \\"babel-eslint\\",
    \\"extends\\": [
      \\"standard\\",
      \\"prettier\\",
      \\"prettier/standard\\"
    ],
    \\"plugins\\": [
      \\"prettier\\"
    ],
    \\"env\\": {
      \\"jest\\": true,
      \\"node\\": true
    },
    \\"rules\\": {
      \\"prettier/prettier\\": [
        \\"error\\",
        {
          \\"semi\\": false,
          \\"singleQuote\\": true,
          \\"trailingComma\\": \\"all\\"
        }
      ]
    }
  },
  \\"eslintIgnore\\": [
    \\"/node_modules\\"
  ],
  \\"devDependencies\\": {
    \\"@commitlint/cli\\": \\"^7.6.1\\",
    \\"@commitlint/config-conventional\\": \\"^7.6.0\\",
    \\"babel-eslint\\": \\"^10.0.1\\",
    \\"commitizen\\": \\"^3.1.1\\",
    \\"cz-conventional-changelog\\": \\"^2.1.0\\",
    \\"eslint\\": \\"^5.16.0\\",
    \\"eslint-config-prettier\\": \\"^4.2.0\\",
    \\"eslint-config-prettier-standard\\": \\"^2.0.0\\",
    \\"eslint-config-standard\\": \\"^12.0.0\\",
    \\"eslint-plugin-import\\": \\"^2.17.2\\",
    \\"eslint-plugin-node\\": \\"^9.0.1\\",
    \\"eslint-plugin-prettier\\": \\"^3.1.0\\",
    \\"eslint-plugin-promise\\": \\"^4.1.1\\",
    \\"eslint-plugin-standard\\": \\"^4.0.0\\",
    \\"husky\\": \\"^2.2.0\\",
    \\"lint-staged\\": \\"^8.1.6\\",
    \\"prettier\\": \\"^1.17.1\\"
  },
  \\"sharec\\": {
    \\"version\\": \\"1.0.0\\"
  }
}"
`;

exports[`commands > install > configuration processing valid cases should merge all configs from target path 6`] = `
Object {
  "bar": Object {
    "bar": Array [
      "foo",
      "bar",
    ],
    "foo": "bar",
  },
  "foo": "bar",
}
`;
